### ISSUE TITLE
Question : is there a way to fetch current url in a script ?

### ISSUE THREAD

**Issue by thibaultdelmas** [next-browser/next#70]
Hi, thanks for the browser, seems tons of fun.

I wanted to start hacking it but I can't seem to find the documentation on how to do what I want.

Basically I want to create a shortcut that when executed takes the url of the page I am currently looking at and paste it to the youtube-dl command line program. So following the example I created a init.lisp script in ~/.config/next/ 

and put in it : 

```
(in-package :next)

(defun dl-video ()
  (uiop:run-program '("youtube-dl" "http:url")))

(define-key *global-map* (kbd "C-d y")#'dl-video)
```

And i'd like to know if there is a global variable (or any other way for that matter) from which I could fetch the current url ( basically the youtube one) and put it at the place of "http:url". 

I've never done any Lisp but since it looked quite straight forward I wanted to have a try at it.

For now this is all I want to do, maybe later add a callback to display the stdout of my program or even test if youtube-dl is installed in the first place and if the link pasted is a youtube url.

If you could help me or point me to the correct documentation it would be very appreciated.

**Comment by lehmacdj** [comment]
You can use this snippet from line 88. We should probably have a helper function that abstracts this.

https://github.com/next-browser/next/blob/master/next/source/document-mode.lisp#L88

**Comment by thibaultdelmas** [comment]
thanks for the quick answer. A simple interface to the document variables would indeed be very nice :).

If I understand correctly I want to use 
(let ((url (interface:web-view-get-url (view (buffer mode)))))

to get the url I want 

Just one little question, which might be tied to the first one, how do I get a reference to the current view from init.lisp ?

**Comment by lehmacdj** [comment]
You can use (view *active-buffer*) to get the view for the active buffer.

**Comment by thibaultdelmas** [comment]
Super cool, that should do the trick. I have some issues tough.

The init.lisp doesn't seem to be loaded when I launch next. ( I am using the .dmg version downloaded from your website ( 0.0.7 ).

If I manually launch it using Ctrl+o, I can use my shortcut but as soon as I do it crashes. Is there a way to get some logs somewhere so I can try to debug it ?

I reduced my init.lisp to the smallest program causing the crash

`(in-package :next)

(defun opencode ()
  (uiop:run-program '("code")))

(define-key *global-map* (kbd "C-d c")
  #'opencode)`
 
if you see anything

**Comment by jmercouris** [comment]
Hi, thank you for your interest in next! I can try to figure out your code tomorrow, but for the mean time, there is a way to get output from next using emacs:

https://github.com/next-browser/next/blob/master/documents/MANUAL.org#slime-with-a-compiled-version-of-next

by connecting to slime you can invoke your own functions, print to output, etc. Thank you for your patience!

**Comment by thibaultdelmas** [comment]
![error](https://user-images.githubusercontent.com/19927521/35556288-1cf034aa-05a2-11e8-9732-a45e1df9200f.png)
I linked a slime to next and apparently the issue is a return code 71. From there I believe the issue come from my code. Strangely enough in the SBCL interpreter I have in a linux vm I tried 

```
(require "asdf")
(asdf:load-system :uiop)
(uiop:run-program '("code"))
```

And it did launch code ( Visual studio Code text editor ).
So I am not really sure what is happening.

For my youtube program, the issue seems simpler.

![error_youtube](https://user-images.githubusercontent.com/19927521/35556849-b89e64d4-05a3-11e8-86c3-ea9c8db8c84a.png)

Basically I think the issue might be either
- the path string is not escaped properly 
or 
- I need to convert the NEXT::VIEW into a simple string
```
(defun dl-video ()
  (uiop:run-program '("youtube-dl" "-o" "\"$HOME/Downloads/%(title)s-%(id)s.%(ext)s\"" ((url (interface:web-view-get-url (view *active-buffer*)))))))

```

**Comment by jmercouris** [comment]
Hi, thank you for your very detailed investigation! Yes, indeed the spec on those dirs changed, but I believe I uploaded a new version of nEXT (also fixes issues of home page path). Can you please confirm that it is working on your machine (the new version)? Perhaps there is some strange caching going on, or something else (my internet was very spotty at the time).

In terms of your issue with "simple-string", this is because you are manually passing in a quoted list like this `'("element 1" "element 2")`, in which case yes, everything must be a simple string. But if you are to use (list "element 1" "element 2") you have a lot more freedom and you can also do something like `(list "element 1" (coerce "element 2" 'simple-string))`. I've provided a brief working example below :)

I was personally having some issues with the pathing as well for the "-o" flag, I'm not sure how to properly escape the strings/format it. I think the best way to figure out that is to construct an example of what you'd like that string to look like in a normal terminal, and then try to recreate that string in a CL REPL. Evaluating a string in the REPL like `"fish"` will immediately return `"fish"`, so you can see how your string is being interpreted.

```
(defun youtube-dl-current-page ()
  (let ((url (interface:web-view-get-url (view *active-buffer*))))
    (uiop:run-program
     (list "youtube-dl" url))))

(define-key *global-map* (kbd "C-c d")
  'youtube-dl-current-page)
```

Anyways, thank you for your patience! Also, thank you for your interest, if you have any more questions, please do not hesitate to join the IRC channel or post new issues!

Thanks,

-John

**Comment by jmercouris** [comment]
What happens when you execute the following:

`(uiop:run-program "echo $PATH" :output t)`

does it show a path in which youtube-dl is also present?

**Comment by thibaultdelmas** [comment]
I am sorry, but I think you broke your last build.
<img width="565" alt="capture d ecran 2018-01-31 a 15 36 20" src="https://user-images.githubusercontent.com/19927521/35628615-bd5558b2-069c-11e8-8948-00e5381c2ee7.png">
<img width="568" alt="capture d ecran 2018-01-31 a 15 34 08" src="https://user-images.githubusercontent.com/19927521/35628616-bd8b10b0-069c-11e8-80a5-9cd2a9f38d3e.png">

I don't know if you look at issues through mail or github, but I edited my last comment a few times. Indeed I think for some reason the PATH is not set through this uiop call. Maybe in a async call ( with launch-program instead, it will be set ).
And since I uninstalled / reinstalled  with the new DMG I can't try for now.

**Comment by jmercouris** [comment]
Huh yeah, that is a problem, it should not be trying to load a dir on my machine :(

**Comment by jmercouris** [comment]
@thibaultdelmas 

I've since fixed the issue, and uploaded a new binary, please let me know if this works for you :)

you'll find a new macro `deferredvar` which is like `defvar` but contains values which are only evaluated at "start" time, so they are not embedded in the lisp image if they need to be calculated based on a user's current environment.

**Comment by thibaultdelmas** [comment]
Well, for some odd reason, if I run next from source in my linux vm. My script works. But on the compiled version, i get permission denied for directory creation from the youtube-dl program I am trying to launch. This seems to be tied to the $PATH issue. Whoami returns nothing. (whereas it returns my username when runned from source). Maybe I will find something from uiop documentation on this behavior and how to set a correct bash environnement. I believe nevertheless it is time to close this issue since the original question was answered : 

To fetch current url from init script, one has to use : 

`(let ((url (interface:web-view-get-url (view *active-buffer*)))))`
